realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp

```diff
@@ -24,42 +24,56 @@
 #include <realm/group_shared.hpp>
 #include <realm/sync/history.hpp>
 #include <realm/sync/client.hpp>
+#include <object-store/src/sync/impl/sync_client.hpp>
 
 #include "objectserver_shared.hpp"
 
 #include "io_realm_SyncManager.h"
 
-#include "jni_util/log.hpp"
-#include "jni_util/jni_utils.hpp"
-#include "sync/sync_manager.hpp"
-#include "sync/sync_user.hpp"
-#include "util.hpp"
-
 using namespace realm;
 using namespace realm::sync;
 using namespace realm::jni_util;
 
 std::unique_ptr<Client> sync_client;
 
+struct AndroidClientListener : public realm::BindingCallbackThreadObserver {
+
+    void did_create_thread() override {
+        realm::jni_util::Log::d("did_create_thread");
+        // Attach the sync client thread to the JVM so errors can be returned properly
+        realm::jni_util::JniUtils::get_env(true);
+    }
+
+    void will_destroy_thread() override {
+        realm::jni_util::Log::d("will_destroy_thread");
+        // Failing to detach the JVM before closing the thread will crash on ART
+        realm::jni_util::JniUtils::detach_current_thread();
+    }
+} s_client_thread_listener;
+
 JNIEXPORT void JNICALL Java_io_realm_SyncManager_nativeInitializeSyncClient
-    (JNIEnv *env, jclass)
+    (JNIEnv* env, jclass)
 {
     TR_ENTER()
     if (sync_client) return;
 
     try {
+        // Setup SyncManager
+        g_binding_callback_thread_observer = &s_client_thread_listener;
```

Comment:
yes it's a global used in the ObjectStore https://github.com/realm/realm-object-store/blob/14c2c7e7038850302f60f6fa1e36a22bceb3ab94/src/sync/impl/sync_client.hpp#L43